File: lib_mvp/src/main/java/com/kasiengao/mvp/java/BasePresenterActivity.java
package com.kasiengao.mvp.java;

import android.content.Context;

import androidx.lifecycle.LifecycleOwner;

/**
 * @ClassName: BasePresenterActivity
 * @Author: KaiSenGao
 * @CreateDate: 2020/3/26 16:00
 * @Description: 带有Presenter 的 Activity基类
 */
public abstract class BasePresenterActivity<Presenter extends BaseContract.BasePresenter> extends BaseToolbarActivity implements BaseContract.BaseView {

    protected Presenter mPresenter;

    @Override
    @SuppressWarnings("unchecked")
    protected void initBefore() {
        super.initBefore();

        this.mPresenter = initPresenter();
        if (this.mPresenter != null) {
            this.mPresenter.attachView(this);
        }
    }

    /**
     * 初始化Presenter
     *
     * @return Presenter
     */
    public abstract Presenter initPresenter();

    /**
     * 显示一个Toast
     *
     * @param message Toast信息
     */
    @Override
    public void showToast(String message) {
        super.showLongSafe(message);
    }

    @Override
    public Context getContext() {
        return this;
    }

    @Override
    public LifecycleOwner getLifecycleOwner() {
        return this;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        if (this.mPresenter != null) {
            this.mPresenter.detachView();
        }
    }
}

File: lib_mvp/src/main/java/com/kasiengao/mvp/kt/BasePresenterActivity.kt
package com.kasiengao.mvp.kt

import android.content.Context
import androidx.lifecycle.LifecycleOwner


/**
 * @ClassName: BasePresenterActivity
 * @Author: KaiSenGao
 * @CreateDate: 2020/3/26 17:04
 * @Description: 带有Presenter 的 Activity基类
 */
abstract class BasePresenterActivity<Presenter : BaseContract.BasePresenter<*>> :
    BaseToolbarActivity(), BaseContract.BaseView {

    protected var mPresenter: Presenter? = null

    override fun initBefore() {
        super.initBefore()

        this.mPresenter = initPresenter()
        @Suppress("UNREACHABLE_CODE", "CAST_NEVER_SUCCEEDS")
        this.mPresenter?.attachView(this as Nothing)
    }

    /**
     * 初始化Presenter
     *
     * @return Presenter
     */
    abstract fun initPresenter(): Presenter

    override val context: Context?
        get() = this

    override val lifecycleOwner: LifecycleOwner?
        get() = this

    override fun onDestroy() {
        super.onDestroy()

        this.mPresenter?.detachView()
    }
}